gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java

```diff
@@ -0,0 +1,154 @@
+   * @throws InvalidProtocolBufferException if failed to serialize the protobuf message to JSON
+   *     format
+   */
+  String toJson(RequestT message) {
+    try {
+      return JsonFormat.printer().print(message);
+    } catch (InvalidProtocolBufferException e) {
+      throw new RestSerializationException("Failed to serialize message to JSON", e);
+    }
+  }
+
+  /**
+   * Deserializes a {@code message} from an input stream to a protobuf message.
+   *
+   * @param message the input stream with a JSON-encoded message in it
+   * @param builder an empty builder for the specific {@code RequestT} message to serialize
+   * @throws RestSerializationException if failed to deserialize a protobuf message from the JSON
+   *     stream
+   */
+  @SuppressWarnings("unchecked")
+  RequestT fromJson(InputStream message, Message.Builder builder) {
+    try (Reader json = new InputStreamReader(message)) {
+      JsonFormat.parser().ignoringUnknownFields().merge(json, builder);
+      return (RequestT) builder.build();
+    } catch (IOException e) {
+      throw new RestSerializationException("Failed to parse response message", e);
+    }
+  }
+
+  /**
+   * Puts a message field in {@code fields} map which will be used to populate URL path of a
+   * request.
+   *
+   * @param fields a map with serialized fields
+   * @param fieldName a field name
+   * @param fieldValue a field value
+   */
+  public void putPathParam(Map<String, String> fields, String fieldName, Object fieldValue) {
+    if (isDefaultValue(fieldName, fieldValue)) {
+      return;
+    }
+    fields.put(fieldName, String.valueOf(fieldValue));
+  }
+
+  /**
+   * Puts a message field in {@code fields} map which will be used to populate query parameters of a
+   * request.
+   *
+   * @param fields a map with serialized fields
+   * @param fieldName a field name
+   * @param fieldValue a field value
+   */
+  public void putQueryParam(Map<String, List<String>> fields, String fieldName, Object fieldValue) {
+    // Avoids empty query parameter
+    if (isDefaultValue(fieldName, fieldValue)) {
+      return;
+    }
+
+    ImmutableList.Builder<String> paramValueList = ImmutableList.builder();
+    if (fieldValue instanceof List<?>) {
+      for (Object fieldValueItem : (List<?>) fieldValue) {
+        paramValueList.add(String.valueOf(fieldValueItem));
+      }
+    } else {
+      paramValueList.add(String.valueOf(fieldValue));
+    }
+
+    fields.put(fieldName, paramValueList.build());
+  }
+
+  /**
+   * Serializes a message to a request body in a form of JSON-encoded string.
+   *
+   * @param fieldName a name of a request message field this message belongs to
+   * @param fieldValue a field value to serialize
+   */
+  public String toBody(String fieldName, RequestT fieldValue) {
+    return toJson(fieldValue);
+  }
+
+  private boolean isDefaultValue(String fieldName, Object fieldValue) {
+    // TODO: Revisit this approach to ensure proper default-value handling as per design.
```

Comment:
any reason not to do this now?

Reply:
The final version most probably will be depending on optional fields in proto itself (i.e. the optiona fields from proto2 back to proto3 requested by Ads and implemented in the recent version of protoc, governed by `--experimental_allow_proto3_optional` flag to protoc). This is not used yet in our generation pipeline, that is why not used here.